### DISCUSSION TITLE
Provide a GitHub Action that translates spotless apply diff into GitHub pull request code suggestion comments

### DISCUSSION THREAD

**Discussion by timtebeek** [diffplug/spotless#1700] [Ideas]
Hi! Couldn't find anything related through search; please correct me if this is not the right place to suggest such a new feature.

*Background*
Lately I've been contributing to Apache Maven, which uses Spotless to enforce formatting; as they've implemented it, the build pipeline breaks and it's up to the developer to apply spotless manually and push a new commit.

*Idea*
Instead I'd hope to see a Spotless GitHub Action developed & applied which takes the diff as created through `spotless:apply` and uses for instance [googleapis/code-suggester](https://github.com/googleapis/code-suggester#Action) to create [code suggestion comments](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request) on the pull request.
This way the developer or reviewer is able to quickly apply all formatting directly from the pull request view, instead of having to checkout a local copy, run spotless again, and commit the resulting changes.

Technically it seems the components are already mostly there, especially when leveraging code-suggester. Having this available as an official GitHub Action could take away some of the friction experienced when using Spotless, especially for incidental contributors.

**Comment by nedtwigg** [comment]
This would be a great idea, it is tracked at

- https://github.com/diffplug/spotless/issues/655

Implementing it is blocked by

- https://github.com/diffplug/spotless/pull/1096
- https://github.com/diffplug/spotless/pull/1097

If someone is eager to build it, the API's in #1096 are stable, you can make a PR against that and it will get merged eventually.

**Comment by timtebeek** [comment]
Ah that's a slightly different approach; would still work but from what I understand it still has some dependencies. I could see the violations format being more broadly applicable, so I understand why you'd nudge towards that existing effort.

With [code-suggester](https://github.com/googleapis/code-suggester#Action) you'd run `spotless:apply` as it is right now, and it'll create the code suggestion comments based on the diff. Could be a quicker way to achieve the same, but understandable if it does not get priority.

**Reply by nedtwigg** [reply]
Interesting, I had not seen code-suggester before. That's a great approach, it doesn't have the dependencies on #1096 and #1097, so I'd be happy to merge something like that now. I'll open an issue referencing this discussion.

**Comment by blacelle** [comment]
This relates with https://github.com/solven-eu/cleanthat#github-app-freepaid which is GitHub App, applying automatically Spotless to unprotected branches, and opening PR for protected branches.

**Reply by nedtwigg** [reply]
Wow, I didn't realize how far along you had already gotten @blacelle! This is great!

If I understand correctly, cleanthat currently provides feedback by opening new PRs. Does it also offer feedback by adding suggested-change diffs to already open PRs?

**Reply by blacelle** [reply]
@nedtwigg this feature was already available in Cleanthat before Spotless integration :).

> cleanthat currently provides feedback by opening new PRs. Does it also offer feedback by adding suggested-change diffs to already open PRs?

There is essentially 2 way of proceeding for now:

1. Either the commit affect a protected branch: then, Cleanthat will suggests changes through a PR. (e.g. https://github.com/solven-eu/cleanthat/pull/493/files)
2. Or the commit affects a not protected branch, used as `head` of a PR to a protected branch: then Cleanthat will apply the changes directly to the branch. (e.g. https://github.com/solven-eu/cleanthat/pull/399/commits/7161270896a5e4c054489bc07b1fa17c3e073583)
3. Other cases are currently discarded (e.g. commit to a PR-less branch)

**Reply by nedtwigg** [reply]
Is a "suggest changes on PRs" option on the roadmap?

**Reply by blacelle** [reply]
Not not yet on the roadmap, but I open https://github.com/solven-eu/cleanthat/issues/552 for the purpose.
